File: macos/ClipRelayMac/Sources/Protocol/MessageCodec.swift
// Binary wire format: length-prefixed message types and encoding/decoding for the L2CAP protocol.

import Foundation
import os

enum MessageType: UInt8 {
    case hello = 0x01
    case welcome = 0x02
    case keyExchange = 0x03
    case keyConfirm = 0x04
    case offer = 0x10
    case accept = 0x11
    case payload = 0x12
    case done = 0x13
    case configUpdate = 0x14
    case reject = 0x15
    case error = 0x16
    case smsSyncRequest = 0x20
    case smsSyncResponse = 0x21
}

struct Message {
    let type: MessageType
    let payload: Data
}

enum ProtocolError: Error, Equatable {
    case incompleteHeader
    case emptyMessage
    case messageTooLarge(Int)
    case incompleteBody
    case unknownType(UInt8)
}

enum MessageCodec {
    static let maxMessageSize = 200_000
    private static let headerSize = 4

    static func encode(_ message: Message) -> Data {
        let messageLength = UInt32(1 + message.payload.count) // type byte + payload
        var data = Data(capacity: headerSize + Int(messageLength))

        // Write length as uint32 big-endian
        var lengthBE = messageLength.bigEndian
        data.append(Data(bytes: &lengthBE, count: 4))

        // Write type byte
        data.append(message.type.rawValue)

        // Write payload
        data.append(message.payload)

        return data
    }

    private static let log = Logger(subsystem: "org.cliprelay", category: "MessageCodec")

    /// Decode a message from a Data buffer at a given offset.
    /// Advances `offset` past the consumed bytes on success.
    /// Skips over messages with unknown types.
    static func decode(from data: Data, offset: inout Int) throws -> Message {
        while true {
            guard data.count - offset >= headerSize else {
                throw ProtocolError.incompleteHeader
            }

            let messageLength = Int(
                UInt32(data[offset]) << 24 |
                UInt32(data[offset + 1]) << 16 |
                UInt32(data[offset + 2]) << 8 |
                UInt32(data[offset + 3])
            )

            offset += headerSize

            if messageLength <= 0 {
                throw ProtocolError.emptyMessage
            }

            if messageLength > maxMessageSize {
                throw ProtocolError.messageTooLarge(messageLength)
            }

            guard data.count - offset >= messageLength else {
                throw ProtocolError.incompleteBody
            }

            let typeByte = data[offset]
            guard let type = MessageType(rawValue: typeByte) else {
                log.warning("Skipping unknown message type 0x\(String(format: "%02x", typeByte))")
                offset += messageLength
                continue
            }

            let payloadStart = offset + 1
            let payloadEnd = offset + messageLength
            let payload = data[payloadStart..<payloadEnd]

            offset = payloadEnd

            return Message(type: type, payload: Data(payload))
        }
    }

    /// Decode a message from an InputStream (blocking read).
    /// Skips over messages with unknown types.
    static func decode(from inputStream: InputStream) throws -> Message {
        while true {
            let headerBytes = try readExactly(from: inputStream, count: headerSize, onFail: .incompleteHeader)

            let messageLength = Int(
                UInt32(headerBytes[0]) << 24 |
                UInt32(headerBytes[1]) << 16 |
                UInt32(headerBytes[2]) << 8 |
                UInt32(headerBytes[3])
            )

            if messageLength <= 0 {
                throw ProtocolError.emptyMessage
            }

            if messageLength > maxMessageSize {
                throw ProtocolError.messageTooLarge(messageLength)
            }

            let body = try readExactly(from: inputStream, count: messageLength, onFail: .incompleteBody)

            let typeByte = body[0]
            guard let type = MessageType(rawValue: typeByte) else {
                log.warning("Skipping unknown message type 0x\(String(format: "%02x", typeByte))")
                continue
            }

            let payload = messageLength > 1 ? Data(body[1...]) : Data()

            return Message(type: type, payload: payload)
        }
    }

    private static func readExactly(from stream: InputStream, count: Int, onFail: ProtocolError) throws -> [UInt8] {
        var buffer = [UInt8](repeating: 0, count: count)
        var totalRead = 0

        while totalRead < count {
            let read = stream.read(&buffer + totalRead, maxLength: count - totalRead)
            if read <= 0 {
                throw onFail
            }
            totalRead += read
        }

        return buffer
    }
}

File: android/app/src/main/java/org/cliprelay/protocol/MessageCodec.kt
package org.cliprelay.protocol

// Binary wire format: length-prefixed message types and encoding/decoding for the L2CAP protocol.

import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.logging.Logger

enum class MessageType(val byte: Byte) {
    HELLO(0x01),
    WELCOME(0x02),
    KEY_EXCHANGE(0x03),
    KEY_CONFIRM(0x04),
    OFFER(0x10),
    ACCEPT(0x11),
    PAYLOAD(0x12),
    DONE(0x13),
    CONFIG_UPDATE(0x14),
    REJECT(0x15),
    ERROR(0x16);

    companion object {
        fun fromByte(b: Byte): MessageType =
            fromByteOrNull(b)
                ?: throw ProtocolException("Unknown message type: 0x${b.toUByte().toString(16).padStart(2, '0')}")

        fun fromByteOrNull(b: Byte): MessageType? =
            entries.firstOrNull { it.byte == b }
    }
}

data class Message(val type: MessageType, val payload: ByteArray) {
    override fun equals(other: Any?): Boolean =
        other is Message && type == other.type && payload.contentEquals(other.payload)

    override fun hashCode(): Int =
        31 * type.hashCode() + payload.contentHashCode()
}

open class ProtocolException(message: String) : Exception(message)
class VersionMismatchException(val version: Int) : ProtocolException("Unsupported protocol version: $version")

object MessageCodec {
    const val MAX_MESSAGE_SIZE = 200_000
    private const val HEADER_SIZE = 4
    private val log = Logger.getLogger("MessageCodec")

    fun encode(message: Message): ByteArray {
        val messageLength = 1 + message.payload.size // type byte + payload
        val buffer = ByteBuffer.allocate(HEADER_SIZE + messageLength)
        buffer.order(ByteOrder.BIG_ENDIAN)
        buffer.putInt(messageLength)
        buffer.put(message.type.byte)
        buffer.put(message.payload)
        return buffer.array()
    }

    fun decode(input: InputStream): Message {
        while (true) {
            val headerBytes = readExactly(input, HEADER_SIZE, "Incomplete header")

            val messageLength = ByteBuffer.wrap(headerBytes).order(ByteOrder.BIG_ENDIAN).int

            if (messageLength <= 0) {
                throw ProtocolException("Empty message: length is $messageLength")
            }

            if (messageLength > MAX_MESSAGE_SIZE) {
                throw ProtocolException("Message too large: $messageLength bytes exceeds maximum of $MAX_MESSAGE_SIZE")
            }

            val body = readExactly(input, messageLength, "Incomplete body")

            val typeByte = body[0]
            val type = MessageType.fromByteOrNull(typeByte)
            if (type == null) {
                log.warning("Skipping unknown message type: 0x${typeByte.toUByte().toString(16).padStart(2, '0')}")
                continue
            }
            val payload = body.copyOfRange(1, body.size)

            return Message(type, payload)
        }
    }

    fun write(output: OutputStream, message: Message) {
        output.write(encode(message))
        output.flush()
    }

    private fun readExactly(input: InputStream, count: Int, onFail: String): ByteArray {
        val buffer = ByteArray(count)
        var offset = 0
        while (offset < count) {
            val read = input.read(buffer, offset, count - offset)
            if (read == -1) {
                throw ProtocolException("$onFail: expected $count bytes, got $offset")
            }
            offset += read
        }
        return buffer
    }
}
